### ISSUE TITLE
Wrong duration calculation

### ISSUE THREAD

**Issue by stueja** [meeting-room-booking-system/mrbs-code#3628]
**Describe the bug**
MRBS is set to "periods" with 1 period. When booking a room with an end date at the end of the month of the start date or with an end date in the next month, the duration is negative.

**To Reproduce**
1. set up MRBS to use periods. Delete the second period so that there is only one period.
2. In the months view, click on a room/day cell
3. In the day view, click on a room/period cell
4. Enter a description for the booking, set the start date to e. g. March 15 and the end date to March 31.
5. click save.
6. in the day view, click on the description
7. in the edit page, watch the duration. (-672 days in this example)

**Expected behavior**
The day calculation should also be correct when bookings roll over into the next month.

**Screenshots**
![edit](https://github.com/meeting-room-booking-system/mrbs-code/assets/9169222/1b36f9e5-84d5-48b3-9ce8-823f46000f74)
![rooms](https://github.com/meeting-room-booking-system/mrbs-code/assets/9169222/fad97146-f1bd-48a2-8ffd-4eb42302ab9a)

**Browser details (please complete the following information):**
 - OS: Arch LInux, Kernel 6.7.8
 - Browser Firefox
 - Version 123.0
 - 
**Server details (please complete the following information):**
 - MRBS version 1.11.4
 - Web server: nginx 1.22.1
 - OS: Ubuntu 22.04
 - PHP version: 8.1.20
 - Database used: mariadb 11.2.3

**Additional context**
I believe the error is somewhere in the recursion in `get_periods_interval`. MWE:
```php
<?php
function get_period_names() : array
{
  static $period_names = null;


  $period_names = array();

  $period_names[0] = ["ganzer Tag"];
  return $period_names;
}


// Gets the interval in periods for a booking with $start_time and $end_time
// Takes account of DST
function get_period_interval(int $start_time, int $end_time) : int
{
  global $periods;

  $periods_per_day = count($periods);

  // Need to use the MRBS version of DateTime to get round a bug in modify()
  // in PHP before 5.3.6.  As we are in the MRBS namespace we will get the
  // MRBS version.
  $startDate = new DateTime();
  $startDate->setTimestamp($start_time);
  echo "startDate: " . $startDate->format('Y-m-d H:i:s') . "\n";
  $endDate = new DateTime();
  $endDate->setTimestamp($end_time);
  echo "endDate: " . $endDate->format('Y-m-d H:i:s') . "\n";

  // Set both dates to noon so that we can compare them and get an integral
  // number of days difference.  Noon also happens to be when periods start,
  // so will be useful in a moment.
  $startDate->modify('12:00');
  $endDate->modify('12:00');
  echo "startDate modified: " . $startDate->format('Y-m-d H:i:s') . "\n";
  echo "endDate modified: " . $endDate->format('Y-m-d H:i:s') . "\n";

  // Calculate the difference in days
  $interval = $startDate->diff($endDate);
  echo "interval: " . $interval->format('%a') . "\n";
  $interval_days = $interval->format('%a');
  echo "interval_days: $interval_days\n";

  if ($interval_days == 0)
  {
    // If the interval starts and ends on the same day, then we just calculate the number
    // of periods by calculating the number of minutes between the start and end times.
    $result = ($end_time - $start_time)/60;
  }
  else
  {
    // Otherwise we calculate the number of periods on the first day
    $startDate->add(new DateInterval('PT' . $periods_per_day . 'M'));
    echo "startDate add new dateinterval: " . $startDate->format('Y-m-d H:i:s') . "\n";
    $result = get_period_interval($start_time, $startDate->getTimestamp());
    // Add in the number of whole days worth of periods in between
    $result += ($interval_days - 1) * $periods_per_day;
    // And add in the number of periods on the last day
    $result += get_period_interval($endDate->getTimestamp(), $end_time);
  }

  return (int) $result;
}

function get_duration(int $start_time, int $end_time, bool $enable_periods, int $area_id, bool $translate=true) : array
{
  $result = array();

  $period_names = get_period_names();

  if ($enable_periods)
  {
    $periods_per_day = count($period_names[$area_id]);
    echo "periods_per_day: $periods_per_day\n";
    $n_periods = get_period_interval($start_time, $end_time);  // this handles DST
    echo "n_periods: $n_periods\n";
    echo "n_periods % periods_per_day: ";
    echo ($n_periods % $periods_per_day);
    if (($n_periods % $periods_per_day) == 0)
    {
      $result['duration'] =  intval($n_periods/$periods_per_day);
      $result['dur_units'] = ($translate) ? get_vocab('days') : 'days';
    }
    else
    {
      $result['duration'] = $n_periods;
      $result['dur_units'] = ($translate) ? get_vocab('periods') : 'periods';
    }
  }
  else
  {
    $result['duration'] = $end_time - $start_time;
    // Need to make DST correct in opposite direction to entry creation
    // so that user see what he expects to see
    $result['duration'] -= cross_dst($start_time, $end_time);
    toTimeString($result['duration'], $result['dur_units'], $translate);
  }
  echo "result: $result\n";
  return $result;
}

    // We will translate the units later
    $data['start_time'] = 1711364400;
    $data['end_time'] = 1711922760;
    $enable_periods = true;
    $data['area_id'] = 0;
    $periods[] = "ganzer Tag";
    $d = get_duration($data['start_time'], $data['end_time'], $enable_periods, $data['area_id'], false);
    $data['duration'] = $d['duration'];
    echo "data duration: \n";
    $data['dur_units'] = $d['dur_units'];
    echo "data units: \n";
```
Output:
```
periods_per_day: 1
startDate: 2024-03-25 11:00:00
endDate: 2024-03-31 22:06:00
startDate modified: 2024-03-25 12:00:00
endDate modified: 2024-03-31 12:00:00
interval: 6
interval_days: 6
startDate add new dateinterval: 2024-03-25 12:01:00
startDate: 2024-03-25 11:00:00
endDate: 2024-03-25 12:01:00
startDate modified: 2024-03-25 12:00:00
endDate modified: 2024-03-25 12:00:00
interval: 0
interval_days: 0
startDate: 2024-03-31 12:00:00
endDate: 2024-03-31 22:06:00
startDate modified: 2024-03-31 12:00:00
endDate modified: 2024-03-31 12:00:00
interval: 0
interval_days: 0
n_periods: 672
n_periods % periods_per_day: 0
Warning: Array to string conversion in /home/user/scripts/code.php on line 100
result: Array
data duration: 
data units: 
```

**Comment by campbell-m** [comment]
Thanks for spotting this.  I think that it's actually a problem with the times that are recorded in the database when the booking crosses a DST boundary.  I'm working on a fix.

**Comment by stueja** [comment]
That looks phantastic on first glance. For me personally, a fix for existing bookings is not necessary. There were only 4 so far, and I simply recreated them by hand. But again, this is just me.
Thanks so much for the quick fix!

**Comment by campbell-m** [comment]
I've now fixed the handling of existing bookings in [6eb686a](https://github.com/meeting-room-booking-system/mrbs-code/commit/6eb686adc7b9da15b86e16312da8f6297c33058a).
